File: Algorithm.java
import java.util.*;
import java.util.stream.Stream;
import java.io.*;

public class Algorithm {
    private ArrayList<ArrayList<String>> Trans;
    private ArrayList<ArrayList<String>> Pq;
    private ArrayList<ArrayList<String>> Util;
    private ArrayList<Integer> uT;
    
    private ArrayList<String> Ikeep;
    private ArrayList<String> profit;
    
    public Algorithm() {
        Trans = new ArrayList<>();
        Pq = new ArrayList<>();
        Util = new ArrayList<>();
        uT = new ArrayList<>();
        Ikeep = new ArrayList<>();
        profit = new ArrayList<>();
    }

    public void readData(String path1, String path2) {
        try {
            File file = new File(path1);
            Scanner sc = new Scanner(file);

            while (sc.hasNextLine()) {
                String line = sc.nextLine();
                String[] parts = line.split("\\|");
                Trans.add(new ArrayList<String>(Arrays.asList(parts[0].split(","))));
                Pq.add(new ArrayList<String>(Arrays.asList(parts[1].split(","))));
                Util.add(new ArrayList<String>(Arrays.asList(parts[2].split(","))));
                
                uT.add(Integer.parseInt(parts[3]));
            }
            sc.close();

            file = new File(path2);
            sc = new Scanner(file);
            String[] K = sc.nextLine().split(",");
            for (String k: K ) {
                Ikeep.add(k);
            }
            
            String[] P = sc.nextLine().split(",");
            for (String p: P) {
                profit.add(p);
            }
            
            sc.close();
            
        } 
        catch (FileNotFoundException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }

    private boolean isSubset(ArrayList<String> parent, ArrayList<String> child) {
        Set<String> p = new HashSet<>(parent);
        return p.containsAll(child);
    }

    public HashMap<ArrayList<String>, Integer> SUP_Item(ArrayList<ArrayList<String>> itemSets) {
        HashMap<ArrayList<String>, Integer> result = new HashMap<>();

        for (ArrayList<String> iSet: itemSets) {
            int counter = 0;
            for (ArrayList<String> tr: Trans) {
                if (isSubset(tr, iSet)) {
                    counter++;
                }
                result.put(iSet, counter);
            }
        }
        return result;
    }

    public int SUP_Items2(ArrayList<String> iSet) {
        int counter = 0;
        for (ArrayList<String> tr: Trans) {
            if (isSubset(tr, iSet)) {
                counter++;
            }
        }
        return counter;
    }

    public ArrayList<Integer> TWU_Items(ArrayList<ArrayList<String>> itemSets) {
        ArrayList<Integer> result = new ArrayList<>();
        
        for (ArrayList<String> iset: itemSets) {
            int s = 0;
            for (int i = 0; i < Trans.size(); i++) {
                if (isSubset(Trans.get(i), iset)) {
                    s += uT.get(i);
                }
            }
            result.add(s);
        }
        return result;
    }

    public int TWU_Items2(ArrayList<String> iSet) {
        int s = 0;
        for (int i = 0; i < Trans.size(); i++) {
            if (isSubset(Trans.get(i), iSet)) {
                s += uT.get(i);
            }
        }
        return s;
    }

    public HashMap<String, Integer> U() {
        HashMap<String, Integer> result = new HashMap<>();
        
        for (int i = 0; i < Trans.size(); i++) {
            for (int j = 0; j < Trans.get(i).size(); j++) {
                String key = Trans.get(i).get(j);
                if (!result.containsKey(key)) {
                    result.put(key, Integer.parseInt(Util.get(i).get(j)));
                }
                else {
                    result.put(key, result.get(key) + Integer.parseInt(Util.get(i).get(j)));
                }
            }   
        }
        return result;
    }

    public boolean isGreater(String x, ArrayList<String> X) {
        ArrayList<String> t1 = new ArrayList<>();
        t1.add(x);
        int a = SUP_Items2(t1);
        int counter = 0;

        for (String element: X) {
            ArrayList<String> t2 = new ArrayList<>();
            t2.add(element);
            int b = SUP_Items2(t2);
            if (a < b || (a == b && (int)(x.charAt(0)) < (int)(element.charAt(0)))) {
                counter++;
            }
        }
        if (counter == X.size()) {
            return true;
        }
        return false;
    }

    public boolean isLess(String x, ArrayList<String> X) {
        ArrayList<String> t1 = new ArrayList<>();
        t1.add(x);
        int a = SUP_Items2(t1);
        int counter = 0;

        for (String element: X) {
            ArrayList<String> t2 = new ArrayList<>();
            t2.add(element);
            int b = SUP_Items2(t2);
            if (a > b || (a == b && (int)(x.charAt(0)) > (int)(element.charAt(0)))) {
                counter++;
            }
        }
        if (counter == X.size()) {
            return true;
        }
        return false;
    }


    public void CoHUI_Miner(double minCor, int minUtility, ArrayList<ArrayList<String>> CoHUIs) {
        ArrayList<String> thr0w = new ArrayList<>();
        
        ArrayList<String> clone = (ArrayList<String>) Ikeep.clone();
        
        for (int i = 0; i < clone.size(); i++) {
            ArrayList<String> temp = new ArrayList<>();
            temp.add(clone.get(i));
            if (TWU_Items2(temp) < minUtility) {
                thr0w.add(Ikeep.get(i));
                Ikeep.remove(i);
            }
        }

        for (int i = 0; i < Trans.size(); i++) {
            Set<String> temp = new HashSet<>(Trans.get(i));
            for (String th: thr0w) {
                if (temp.contains(th)) {
                    int j = Trans.get(i).indexOf(th);
                    Trans.get(i).remove(j);
                    Pq.get(i).remove(j);
                    uT.set(i, uT.get(i) - Integer.parseInt(Util.get(i).remove(j)));
                }
            }

            SORT(Trans.get(i), Pq.get(i), Util.get(i));
        }

        HashMap<String, Integer> U = U();
        ArrayList<ArrayList<String>> itemSets = new ArrayList<>();
        Cardinality(Ikeep, 1, itemSets, new ArrayList<String>());
        
        for (String X: Ikeep) {
            ArrayList<String> XList = new ArrayList<>();
            XList.add(X);
            if (U.get(X) >= minUtility) {    
                CoHUIs.add(XList);
                System.out.print(XList + " - ");
                System.out.println(U.get(X));
            }

            int RemainU = 0;
            ArrayList<Integer> pru_Tx = new ArrayList<>();
            
            ArrayList<Integer> uT_Tx = new ArrayList<>();
            ArrayList<ArrayList<Integer>> utils = new ArrayList<>();
            ArrayList<ArrayList<String>> dbProjectX = new ArrayList<>();

            for (int i = 0; i < Trans.size(); i++) {
                int j = 0;
                int uTemp = uT.get(i);

                while (j < Trans.get(i).size() && isGreater(Trans.get(i).get(j), XList)) {
                    uTemp -= Integer.parseInt(Util.get(i).get(j));
                    j++;
                }

                if (j == Trans.get(i).size() || isLess(Trans.get(i).get(j), XList)) {
                    continue;
                }

                else if (j < Trans.get(i).size()) {
                    ArrayList<Integer> util_ = new ArrayList<>();
                    ArrayList<String> Tx = new ArrayList<>();
                    for (int k = j+1; k < Trans.get(i).size(); k++) {
                        Tx.add(Trans.get(i).get(k));
                        util_.add(Integer.parseInt(Util.get(i).get(k)));
                    }
                    utils.add(util_);
                    pru_Tx.add(Integer.parseInt(Util.get(i).get(j)));
                    uT_Tx.add(uTemp);
                    dbProjectX.add(Tx);
                    RemainU += uTemp;
                }
            }

            SearchCoHUI(XList, U.get(X), RemainU, dbProjectX, 2, pru_Tx, utils, minUtility, minCor, uT_Tx, CoHUIs);
        }
        
        
    }

    private void SearchCoHUI(ArrayList<String> X, int Ux, int RUx, ArrayList<ArrayList<String>> dbProjectX, int k, ArrayList<Integer> pru_Tx, ArrayList<ArrayList<Integer>> utils, int minUtility, double minCor, ArrayList<Integer> uT_Tx, ArrayList<ArrayList<String>> CoHUIs) {
        if (k == Ikeep.size())  return;
        ArrayList<ArrayList<String>> itemSets = new ArrayList<>();
        Cardinality(Ikeep, k, itemSets, new ArrayList<>());
       
        for (ArrayList<String> iSet: itemSets) {
            if (isExtends(X, iSet, k)) {
                
                ArrayList<String> lastItem = (ArrayList<String>) iSet.clone();
                // System.out.println(lastItem);

                ArrayList<String> Xx = new ArrayList<>(X);
                if(!X.equals(lastItem)) {
                    for (String item: lastItem) {
                        if (!Xx.contains(item)) {
                            Xx.add(item);
                        }
                    }
                }
                

                ArrayList<ArrayList<String>> newdbProjectX = new ArrayList<>();
                ArrayList<Integer> newuT_Tx = new ArrayList<>();
                ArrayList<Integer> newpru_TX = new ArrayList<>();
                ArrayList<ArrayList<Integer>> newutils = new ArrayList<>();
                int RUx_ = 0;
                int SUPx_ = 0;
                int Ux_ = Ux;
                int ULA = Ux + RUx;

                for (int i = 0; i < dbProjectX.size(); i++) {
                    int j = 0;
                    int uTemp = uT_Tx.get(i);

                    while (j < dbProjectX.get(i).size() && isGreater(dbProjectX.get(i).get(j), lastItem)) {
                        uTemp -= utils.get(i).get(j);
                        j++;
                    }

                    if (j == dbProjectX.get(i).size() || isLess(dbProjectX.get(i).get(j), lastItem)) {
                        Ux_ -= pru_Tx.get(i);
                        ULA -= (pru_Tx.get(i) + uT_Tx.get(i));

                        if (ULA < minUtility) {
                            return;
                        }
                        continue;
                    }
                    else {
                        Ux_ += utils.get(i).get(j);
                        SUPx_++;

                        if (j < dbProjectX.get(i).size()) {
                            ArrayList<Integer> util_ = new ArrayList<>();
                            ArrayList<String> Tx = new ArrayList<>();
                            for (int h = j+1; h < dbProjectX.get(i).size(); h++) {
                                Tx.add(dbProjectX.get(i).get(h));
                                
                                util_.add(utils.get(i).get(h));
                            }
                            newdbProjectX.add(Tx);
                            newutils.add(util_);
                            newpru_TX.add(pru_Tx.get(i) + utils.get(i).get(j));
                            newuT_Tx.add(uTemp);
                            RUx_ += uTemp;
                        }
                    }
                }
                
                if (SUPx_ > 0) {
                    if (Kulc(Xx) >= minCor) {
                        if (Ux_ >= minUtility) {
                            CoHUIs.add(Xx);
                            System.out.print(Xx + " - ");
                            System.out.println(Ux_);
                        }

                        if ((Ux_ + RUx_) >= minUtility) {
                            SearchCoHUI(Xx, Ux_, RUx_, newdbProjectX, k+1, newpru_TX, newutils, minUtility, minCor, newuT_Tx, CoHUIs);
                        }
                    }    
                }
            }
        }
    }

    private double Kulc(ArrayList<String> X_) {
        double counter = 0;
        for (ArrayList<String> T: Trans) {
            if (isSubset(T, X_)) {
                counter++;
            }
        }

        double result = 0;
        for (String x: X_) {
            ArrayList<String> temp = new ArrayList<>();
            temp.add(x);
            result += (double) (counter / SUP_Items2(temp));
        }
        
        return (double) result / X_.size();
    }

    private boolean isExtends(ArrayList<String> s1, ArrayList<String> s2, int k) {
        if (k == 1) return s1.equals(s2);
        
        for (int i = 0; i < k - 1; i++) {
            if (!s1.get(i).equals(s2.get(i))) {
                return false;
            }
        }
        return true;       
    }

    private int charToInt(String s) {
        return (int)(s.charAt(0));
    }

    private String intToChar(int n) {
        return "" + (char) n;
    }

    private void Cardinality(ArrayList<String> S, int k, ArrayList<ArrayList<String>> result, ArrayList<String> x) {
        if (x.size() < k)  {
            int a = (x.size() == 0) ? charToInt(S.get(0)) : charToInt(x.get(x.size() - 1)) + 1;
            int b = charToInt(S.get(S.size() - 1)) + 1;
            ArrayList<Integer> candidates = new ArrayList<>();
            for (int i = a; i < b; i++) {
                candidates.add(i);
            }

            for (Integer c: candidates) {
                ArrayList<String> t = (ArrayList<String>) x.clone();
                t.add(intToChar(c));
                
                if (t.size() == k) {
                    result.add(t);
                }
                Cardinality(S, k, result, t);
            }
        }
    }

    private void SORT(ArrayList<String> trans, ArrayList<String> pq, ArrayList<String> utils) {
        int n = trans.size();
        for (int i = 0; i < n - 1; i++) {
            ArrayList<String> a = new ArrayList<>();
            
            for (int j = i+1; j < n; j++) {
                ArrayList<String> b = new ArrayList<>();
                a.add(trans.get(i));
                b.add(trans.get(j));
                if (SUP_Items2(a) > SUP_Items2(b)) {
                    Collections.swap(trans, i, j);
                    Collections.swap(pq, i, j);
                    Collections.swap(utils, i, j);
                }

                else if (SUP_Items2(a) == SUP_Items2(b)) {
                    if (charToInt(trans.get(i)) > charToInt(trans.get(j))) {
                        Collections.swap(trans, i, j);
                        Collections.swap(pq, i, j);
                        Collections.swap(utils, i, j);
                    }
                }
                
            } 
        }
    }



    public void show() {
        for (int i = 0; i < Trans.size(); i++) {
            String trans = "";
            String pq = "";
            String utils = "";
            for (int j = 0; j < Trans.get(i).size(); j++) {
                trans += Trans.get(i).get(j) + " ";
                pq += Pq.get(i).get(j) + " ";
                utils += Util.get(i).get(j) + " ";
            }
            System.out.println(trans + " | " + pq + " | " + utils + " | " + uT.get(i).toString());
        }
    }

    public static void main(String[] args) {
        Algorithm Al = new Algorithm();
        Al.readData("data.txt", "profit.txt");
        // Al.show();
        ArrayList<ArrayList<String>> CoHUIS = new ArrayList<>();
        Al.CoHUI_Miner(0.52,70, CoHUIS);
        //Al.show();
    }
}

File: Algorithm.py
from functions import *

root = '/Users/nkeyskuo/Documents/WorkSpace/Algorithm/Final/'

class Algorithm:
    def __init__(self):
        self.Trans = [] # Transaction 
        self.Pq = [] # Purchase Quantity
        self.Util = [] # Utility
        self.uT = [] # Transaction Utility

        self.Ikeep = [] # [a,b,c,d,e,...] - all of single item
        self.profit = [] # [4,3,1,2,3,...] = - profit of single item

    def readData(self, path1, path2):
        # input: path1 - string, path2: string
        # output: Trans, Pq, Util, uT, Ikeep, profit
        file = open(path1, 'r')
        docx = file.read().split('\n')
        for row in docx:
            r = row.split('|')
            self.Trans.append(r[0].split(','))
            self.Pq.append([int(i) for i in r[1].split(',')])
            self.Util.append([int(i) for i in r[2].split(',')])
            self.uT.append([int(r[3])])
        
        file = open(path2, 'r')
        docx = file.read().split('\n')
        self.Ikeep = docx[0].split(',')
        self.profit = [int(i) for i in docx[1].split(',')]

    # input: itemSets - list of ItemSet
    # output: occurrences of every itemSet
    def SUP_Items(self, itemSets):
        result = {}

        for iSet in itemSets:
            counter = 0
            for tr in self.Trans:
                if all(element in tr for element in iSet):
                    counter += 1
            result[str(iSet)] = counter 
        return result


    # input: iSet - single ItemSet
    # output: occurrence the iSet
    def SUP_Items2(self, iSet):
        counter = 0
        for tr in self.Trans:
            if all(element in tr for element in iSet):
                    counter += 1
        return counter

    
    # input: itemSets - list of ItemSet
    # output: Transaction Utilities of every itemSet
    def TWU_Items(self, itemSets):
        result = []
        
        for iSet in itemSets:
            s = 0 
            for i in range(len(self.Trans)):
                if all(element in self.Trans[i] for element in iSet):
                    s += self.uT[i][0]
            result.append(s)

        return result

    # input: iSet - single of ItemSet
    # output: Transaction Utility of the iSet
    def TWU_Items2(self, iSet):
        s = 0
        for i in range(len(self.Trans)):
            if all(element in self.Trans[i] for element in iSet):
                s += self.uT[i][0]
        return s

    # input: List of transactions
    # output: Sum of Utilitities of every elements
    def U(self):
        U = {}
        for i in range(len(self.Trans)):
            for j in range(len(self.Trans[i])):
                if U.get(self.Trans[i][j]) is None:
                    U[self.Trans[i][j]] = self.Util[i][j]
                else:
                    U[self.Trans[i][j]] += self.Util[i][j]
        
        return U

    # input: x - an element, X - a set of elements
    # output: return true if X > x else false
    def isGreater(self, x, X):
        a = self.SUP_Items2(x)
        counter = 0
        for element in X:
            b = self.SUP_Items2(element)
            if a < b or (a == b and ord(x) < ord(element)):
                counter += 1
        if counter == len(X):
            return True
        return False

    # input: x - an element, X - a set of elements
    # output: return true if X < x else false
    def isLess(self, x, X):
        a = self.SUP_Items2(x)
        counter = 0
        for element in X:
            b = self.SUP_Items2(element)
            if a > b or (a == b and ord(x) > ord(element)):
                counter += 1
        if counter == len(X):
            return True
        return False

    # input: database, minCor , minUtility
    # output: CoHuis - set of all correlated HUIs
    def CoHUI_Miner(self, minCor, minUtility, CoHUIs):
        throw = []
        
        for i in range(len(deepcopy(self.Ikeep))):
            if self.TWU_Items2(self.Ikeep[i]) < minUtility:
                throw.append(self.Ikeep[i])
                self.Ikeep.pop(i)
        
        for i in range(len(self.Trans)):
            for th in throw:
                if th in self.Trans[i]:
                    j = self.Trans[i].index(th)
                    self.Trans[i].remove(th)
                    self.Pq[i].pop(j)
                    self.uT[i][0] -= self.Util[i].pop(j)
        
            self.SORT(self.Trans[i], self.Pq[i], self.Util[i])

        U = self.U()
        
        itemSets = []
        Cardinality(self.Ikeep, 1, itemSets, [])

        for X in self.Ikeep:
            if U[X] >= minUtility:
                CoHUIs.append([X])
                print([X], end=' - ')
                print(U[X])
                
            RemainU = 0
            pru_Tx = []
            Tx = 0
            uT_Tx = []
            utils = []
            dbProjectX = []
                
            for i in range(len(self.Trans)):
                j = 0
                uTemp = self.uT[i][0]
                
                while j < len(self.Trans[i]) and self.isGreater(self.Trans[i][j], X):
                    uTemp -= self.Util[i][j]
                    j += 1
                
                if j == len(self.Trans[i]) or self.isLess(self.Trans[i][j], X):
                    continue
                elif j < len(self.Trans[i]):
                    Tx = self.Trans[i][j+1:]
                    pru_Tx.append(self.Util[i][j])
                    uT_Tx.append(uTemp)
                    utils.append(self.Util[i][j+1:])
                    dbProjectX.append(Tx)
                    RemainU += uTemp
            
            self.SearchCoHUI([X], U[X], RemainU, dbProjectX, 2, pru_Tx, utils, minUtility, minCor, uT_Tx, CoHUIs)
        
        

    # input: X : prefix itemset; U (X ): utility of X , RU (X ): the remain utility of X , dbProjectX : projected database with X prefix; k : length of items set X .
    # output: itemsets are CoHUIs with X prefix
    def SearchCoHUI(self, X, Ux, RUx, dbProjectX, k, pru_Tx, utils, minUtility,  minCor, uT_Tx, CoHUIs):
        if k == len(self.Ikeep):
            return 
        itemSets = []
        Cardinality(self.Ikeep, k, itemSets, [])
        
        for iSet in itemSets:
            if isExtends(X, iSet, k):
                lastItem = iSet
                
                Xx = X if (lastItem in X or lastItem == X) else X + list(set(lastItem) - set(X))
               
                newdbProjectX = []
                newuT_Tx = []
                newpru_TX = []
                newutils = []
                RUx_ = 0
                SUPx_ = 0
                Ux_ = Ux
                ULA = Ux + RUx

                for i in range(len(dbProjectX)):
                    j = 0
                    uTemp = uT_Tx[i]

                    while j < len(dbProjectX[i]) and self.isGreater(dbProjectX[i][j], lastItem):
                        uTemp -= utils[i][j]
                        j += 1
                    if j == len(dbProjectX[i]) or self.isLess(dbProjectX[i][j], lastItem):
                        Ux_ -= pru_Tx[i]
                        ULA -= (pru_Tx[i] + uT_Tx[i])

                        if ULA < minUtility:
                            return
                        continue
                    else:
                        Ux_ += utils[i][j]
                        SUPx_ += 1

                        if j < len(dbProjectX[i]):
                            newdbProjectX.append(dbProjectX[i][j+1:])
                            newpru_TX.append(pru_Tx[i] + utils[i][j])
                            newuT_Tx.append(uTemp)
                            newutils.append(utils[i][j+1:])
                            RUx_ += uTemp

                if SUPx_ > 0:
                    if self.Kulc(Xx) >= minCor:
                        if Ux_ >= minUtility:
                            CoHUIs.append(Xx)
                            print(Xx, end=' - ')
                            print(Ux_)

                        if Ux_ + RUx_ >= minUtility:
                            self.SearchCoHUI(Xx, Ux_, RUx_, newdbProjectX, k+1, newpru_TX, newutils, minUtility, minCor, newuT_Tx, CoHUIs)
                
    # input: X_ - an ItemSet
    # output: 
    def Kulc(self, X_):
        counter = 0
        for T in self.Trans:
            if all(element in T for element in X_):
                counter += 1
        
        res = 0
        for x in X_:
            res += (counter / self.SUP_Items2(x))

        return res / len(X_)
        


    def show(self):
        combo = []
        Cardinality(self.Ikeep,3, combo, [])
        print(combo)
        # self.CoHUI_Miner(1, 70)
        # for i in range(len(self.Trans)):
            
        #     print(self.Trans[i], end=' | ')
        #     print(self.Pq[i], end=' | ')
        #     print(self.Util[i], end=' | ')
        #     print(self.uT[i])

    # input : database
    # output: ascending sort according to SUP
    def SORT(self, trans, pq, utils):
        n = len(trans)
        for i in range(0,n-1):
            for j in range(i+1, n):
                if self.SUP_Items2(trans[i]) > self.SUP_Items2(trans[j]):
                    swap(trans,i,j)
                    swap(pq,i,j)
                    swap(utils,i,j)
                elif self.SUP_Items2(trans[i]) == self.SUP_Items2(trans[j]):
                    if ord(trans[i]) > ord(trans[j]):
                        swap(trans,i,j)
                        swap(pq,i,j)
                        swap(utils,i,j)
                

if __name__ == '__main__':
    Al = Algorithm()
    Al.readData(root + 'data.txt', root + 'profit.txt')
    CoHUIs = []
    
    Al.CoHUI_Miner(0.52,70, CoHUIs)
    
    # print(CoHUIs)
    
    
        